src/metabase/util/i18n/plural.clj

```diff
@@ -0,0 +1,116 @@
+  grammar for all operators at or above a single level of precedence.
+
+  The `instaparse` README (https://github.com/Engelberg/instaparse) has an example of a parser called `arithmetic`
+  which is essentially a simpler version of this exact same parser. It may help to read and understand that parser
+  first before trying to understand this one."
+  (insta/parser
+   "expr           = <s> maybe-ternary <s> <';'>? <s>
+
+   <maybe-ternary> = ternary | maybe-or
+   ternary         = maybe-or <s> <'?'> <s> maybe-ternary <s> <':'> <s> maybe-ternary
+
+   <maybe-or>      = or-expr | maybe-and
+   or-expr         = maybe-or <s> <'||'> <s> maybe-and
+
+   <maybe-and>     = and-expr | maybe-eq
+   and-expr        = maybe-and <s> <'&&'> <s> maybe-eq
+
+   <maybe-eq>      = eq-expr | neq-expr | maybe-comp
+   eq-expr         = maybe-eq <s> <'=='> <s> maybe-comp
+   neq-expr        = maybe-eq <s> <'!='> <s> maybe-comp
+
+   <maybe-comp>    = lt-expr | lte-expr | gt-expr | gte-expr | maybe-add
+   lt-expr         = maybe-comp <s> <'<'> <s> maybe-add
+   lte-expr        = maybe-comp <s> <'<='> <s> maybe-add
+   gt-expr         = maybe-comp <s> <'>'> <s> maybe-add
+   gte-expr        = maybe-comp <s> <'>='> <s> maybe-add
+
+   <maybe-add>     = add-expr | sub-expr | maybe-mult
+   add-expr        = maybe-add <s> <'+'> <s> maybe-mult
+   sub-expr        = maybe-add <s> <'-'> <s> maybe-mult
+
+   <maybe-mult>    = mult-expr | div-expr | mod-expr | operand
+   mult-expr       = maybe-mult <s> <'*'> <s> operand
+   div-expr        = maybe-mult <s> <'/'> <s> operand
+   mod-expr        = maybe-mult <s> <'%'> <s> operand
+
+   <operand>       = integer | variable | parens
+   <parens>        = <'('> <s> expr <s> <')'>
+   <s>             = <#'\\s+'>*
+   integer         = #'[0-9]+'
+   variable        = 'n'"))
+
+(defn- to-bool
+  "Converts an integer or Boolean to a Boolean to use in a C-style logical operator."
+  [x]
+  (if (integer? x)
+    (if (= x 0) false true)
+    x))
+
+(defn- to-int
+  "Converts an integer or Boolean to an integer to use in a C-style arithmetic operator."
+  [x]
+  (if (boolean? x)
+    (if x 1 0)
+    x))
+
+(defn- op
+  "Converts a Clojure binary function f to a C-style operator that treats Booleans as integers, and returns an integer."
+  [f]
+  (fn [x y] (to-int (f (to-int x) (to-int y)))))
+
+(defn- tag-fns
+  "Functions to use for each tag in the parse tree, when transforming the tree into a single value."
+  [n]
+  {:add-expr  (op +)
+   :sub-expr  (op -)
+   :mult-expr (op *)
+   :div-expr  (op /)
+   :mod-expr  (op mod)
+   :eq-expr   (op =)
+   :neq-expr  (op not=)
+   :gt-expr   (op >)
+   :gte-expr  (op >=)
+   :lt-expr   (op <)
+   :lte-expr  (op <=)
+   :and-expr  #(to-int (and (to-bool %1) (to-bool %2)))
+   :or-expr   #(to-int (or (to-bool %1) (to-bool %2)))
+   :ternary   #(to-int (if (to-bool %1) %2 %3))
+   :integer   #(Integer. ^String %)
+   :variable  (constantly n)
+   :expr      identity})
+
+(def index
```

Comment:
nitty and might be over-engineering:
do you think it's worth it to have a cache for the parse function? 
it seems like an expensive operation that has limited number of inputs, maybe a threshold of 10 is already sufficient?

Reply:
This function is memozied (see the comment for an explanation of the reasoning); is that different from what you're describing?